src/structs/lepton_decoder.rs

```diff
@@ -270,47 +290,44 @@ fn parse_token<R: Read, const ALL_PRESENT: bool>(
     let mut eob_y: u8 = 0;
     let mut num_non_zeros_left_7x7: u8 = num_non_zeros_7x7;
 
+    let best_priors =
+        pt.calc_coefficient_context_7x7_aavg_block::<ALL_PRESENT>(image_data, context);
+
+    let block = context.here_mut(image_data).get_block_mut();
     for zz in 0..49 {
         if num_non_zeros_left_7x7 == 0 {
             break;
         }
 
-        let coord = UNZIGZAG_49[zz];
-        let b_x = coord & 7;
-        let b_y = coord >> 3;
+        let best_prior_bit_length = u16_bit_length(best_priors[zz] as u16);
 
+        let coord = UNZIGZAG_49[zz];
         assert!(
             (coord & 7) > 0 && (coord >> 3) > 0,
             "this does the DC and the lower 7x7 AC"
         );
 
-        let ptcc7x7 = pt.calc_coefficient_context_7x7_aavg::<ALL_PRESENT>(
-            image_data,
-            coord as usize,
-            context,
-            num_non_zeros_left_7x7,
-        );
-
         let coef = model
             .read_coef(
                 bool_reader,
                 pt.get_color_index(),
                 coord.into(),
                 zz.into(),
-                ptcc7x7.num_non_zeros_bin as usize,
-                ptcc7x7.best_prior_bit_len as usize,
+                ProbabilityTables::num_non_zeros_to_bin(num_non_zeros_left_7x7) as usize,
+                best_prior_bit_length as usize,
             )
             .context(here!())?;
 
         if coef != 0 {
+            let b_x = coord & 7;
+            let b_y = coord >> 3;
+
             eob_x = cmp::max(eob_x, b_x);
             eob_y = cmp::max(eob_y, b_y);
             num_non_zeros_left_7x7 -= 1;
         }
 
-        context
-            .here_mut(image_data)
-            .set_coefficient(zz + ALIGNED_BLOCK_INDEX_AC_7X7_INDEX, coef);
+        block[zz as usize + ALIGNED_BLOCK_INDEX_AC_7X7_INDEX] = coef;
```

Comment:
yes, good idea: better to do the pointer math here than the more complex call here.